### PR TITLE
Enable test algorithm

### PR DESCRIPTION
**Pull request by ehebert**
@jbredeche these patches enable more of test_algorithm, by using the old trade data to create DataPortal's, up to the point of discovering a bug in when orders are filled in test_volshare_slippage

### CODE REVIEW

File: tests/utils/test_utils.py [hunk 1]
```diff
@@ -62,3 +62,72 @@ def create_data_portal(env, tempdir, sim_params, sids):
             sim_params=sim_params,
             asset_finder=env.asset_finder
         )
+
+
+def create_data_portal_from_trade_history(
```
Comment (jbredeche) [inline tests/utils/test_utils.py]: can this method have the params in the same order as ```create_data_portal```?

ie:
```
def create_data_portal_from_trade_history(env, tempdir, sim_params, trades_by_sid)
```

File: zipline/finance/slippage.py [hunk 2]
```diff
@@ -314,10 +314,13 @@ def __init__(self, spread=0.0):
         self.spread = spread
 
     def process_order(self, event, order, dt):
```
Comment (ssanderson) [inline zipline/finance/slippage.py]: It looks like this no longer needs to be passed the event, which raises the question: **anything** should ever get passed an `Event` object anymore?  It seems like the answer should be "No"?

File: zipline/finance/slippage.py [hunk 3]
```diff
@@ -314,10 +314,13 @@ def __init__(self, spread=0.0):
...
+
+        price = self.data_portal.get_current_price_data(order.sid, 'close')
```
Comment (jbredeche) [inline zipline/finance/slippage.py]: where is ```self.data_portal``` assigned?

File: zipline/gens/tradesimulation.py [hunk 4]
```diff
@@ -188,6 +188,10 @@ def inner_loop(dt_to_use):
                     perf_tracker_benchmark_returns[trading_day] = 0.001
                     # use the last dt as market close
                     yield self.get_message(trading_day)
+
+                    self.algo.portfolio_needs_update = True
```
Comment (jbredeche) [inline zipline/gens/tradesimulation.py]: why the change?

File: zipline/algorithm.py [hunk 5]
```diff
@@ -445,6 +440,10 @@ def _create_generator(self, sim_params, source_filter=None):
                 data_portal=self.data_portal
             )
 
+        if not self.initialized:
```
Comment (jbredeche) [inline zipline/algorithm.py]: why push back initialization to here?

File: zipline/finance/slippage.py (hunk 2 above)
Comment (ehebert) [inline zipline/finance/slippage.py]: I think when the dust settles, that will be the case.

File: zipline/gens/tradesimulation.py (hunk 4 above)
Comment (ehebert) [inline zipline/gens/tradesimulation.py]: IIRC, found this when a unit test tried to access account, and the `needs_update` flag was initially False... the `_needs_update` fields may subject to removal, since some of the bottlenecks they were addressing may go away.

File: zipline/finance/slippage.py (hunk 3 above)
Comment (ehebert) [inline zipline/finance/slippage.py]: It is 'hotwired' in in tradesimulation, which was pre-existing to this patch. (VolumeShare also uses the data_portal.)

We should pass it explicitly.

File: zipline/algorithm.py (hunk 5 above)
Comment (ehebert) [inline zipline/algorithm.py]: I can rewind the history a bit to find the exact algo, but IIRC it was because an algorithm needed access to a perf_tracker during initialize.

Note that between the lazy-mainline branch and master we made this change:

```
-        # Build a perf_tracker
-        self.perf_tracker = PerformanceTracker(sim_params=self.sim_params,
-                                               env=self.trading_environment)
-
+        self.perf_tracker = None
         # Pull in the environment's new AssetFinder for quick reference
```
